core/src/main/scala/org/apache/spark/sql/execution/row/RowDeleteExec.scala

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2017 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package org.apache.spark.sql.execution.row
+
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode, ExpressionCanonicalizer}
+import org.apache.spark.sql.catalyst.expressions.{Attribute, BindReferences}
+import org.apache.spark.sql.execution.{SparkPlan, TableMutationExec}
+import org.apache.spark.sql.sources.ConnectionProperties
+import org.apache.spark.sql.store.CodeGeneration
+import org.apache.spark.sql.types.{StructField, StructType}
+
+/**
+ * Generated code plan for deletes in a row table.
+ */
+case class RowDeleteExec(child: SparkPlan, resolvedName: String,
+    tableSchema: StructType, keyColumns: Seq[Attribute],
+    connProps: ConnectionProperties, onExecutor: Boolean)
+    extends TableMutationExec(forUpdate = false) {
+
+  @transient private var result: String = _
+  @transient private var stmt: String = _
+  @transient private var batchCount: String = _
+
+  override protected def doProduce(ctx: CodegenContext): String = {
+    val (initCode, commitCode, endCode) = connectionCodes(ctx)
+    result = ctx.freshName("result")
+    ctx.addMutableState("long", result, s"$result = -1L;")
+    stmt = ctx.freshName("statement")
+    batchCount = ctx.freshName("batchCount")
+    val childProduce = doChildProduce(ctx)
+    s"""
+       |if ($result >= 0L) return;
+       |$initCode
+       |try {
+       |  final java.sql.PreparedStatement $stmt = $connTerm.prepareStatement(
+       |    "${s"DELETE FROM $resolvedName WHERE $whereClauseString"}");
```

Comment:
To add to the "deleteFrom" comment:

You are probably confused between two completely different use-cases. One is where one would want to say "delete all rows from target table matching given input obtained from some source". This is deleteFrom which is more for CDC-like cases. The more common cases are different:

"update all rows in the table matching ABC criteria to change to XYZ"

"delete all rows from a table matching ABC criteria"

I don't see any clean way to express 1 using writer-like APIs. One could possibly express 2 as: snappy.table("table1").filter(ABC)....write.deleteFrom("table1") but this is:

a) quite convoluted

b) is in-congruent with the corresponding SQL: delete from table1 where ABC

c) is very inefficient when table has no primary key

d) is inefficient even if there is a primary key since the corresponding deletes are row-wise that has to check all the columns of the table like: "delete from table1 where col1=? and col2=? ..."